<div style="border:solid green 2px; padding: 20px">
<b>Hello, Noa!</b>
    
Grigoriy is here again.  

You can find my comments in <font color='green'>green</font> or <font color='red'>red</font> boxes like this:

<div class="alert alert-success">
<b>Reviewer's comment:</b> if everything is done succesfully
</div>

<div class="alert alert-danger">
<b>Reviewer's comment:</b> if the block requires some corrections. Work can't be accepted with the red comments.
</div>
    
Let's work on the project in dialogue: if you change something in the project or respond to my comments, write about it. It will be easier for me to track the changes if you highlight your comments:
    
<div class="alert alert-info"> <b>Student's comment:</b> For example like this.</div>    
    
      
<div>

### <font color='orange'>General feedback</font>
Thank you for sending your project.  In general - good job, just pay attention to the second task. 

**Update**:<br>
Thank you for the update! Please, take the second glance to the second task.

**Update**:<br>
Well done, your project has been accepted!

## SQL Project
#### Description
The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.
You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

## Table of Contents <a class="anchor" id="table_of_contents"></a>

### [Step 0. Describe the goals of the research](#chapter0)
Instructions for completing the task:
- Describe the goals of the study.
- Study the tables (print the first rows).
- Make an SQL query for each of the tasks.
- Output the results of each query in the Notebook.
- Describe your conclusions for each of the tasks.

### [Step 1. Tasks](#chapter1)
1. Find the number of books released after January 1, 2000.
2. Find the number of user reviews and the average rating for each book.
3. Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
4. Identify the author with the highest average book rating: look only at books with at least 50 ratings.
5. Find the average number of text reviews among users who rated more than 50 books.

<div class="alert alert-success">
<b>Reviewer's comment: </b> Nice introduction!
</div>

### [Step 0. Downloading the data and investigating it](#chapter0)

In [1]:
# importing the necessary libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
# studying the tables
# defining the function to display the head of the tables
def func_display(x):
    x = pd.io.sql.read_sql(x, con = engine)
    display(x.head())

# studying 'books' table
books = '''
SELECT * FROM
    books
'''
func_display(books)

# studying 'authors' table
authors = '''
SELECT * FROM
    authors
'''
func_display(authors)

# studying 'publishers' table
publishers = '''
SELECT * FROM
    publishers
'''
func_display(publishers)

# studying 'ratings' table
ratings = '''
SELECT * FROM
    ratings
'''
func_display(ratings)

# studying 'reviews' table
reviews = '''
SELECT * FROM
    reviews
'''
func_display(reviews)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-success">
<b>Reviewer's comment: </b> Well done, let's move on to the task part.
</div>

### [Step 1. Tasks](#chapter1)
[Go back to the Table of Contents](#table_of_contents)

#### 1. Find the number of books released after January 1, 2000.

In [3]:
query_0 = f'''
SELECT COUNT(book_id) FROM books AS books_number
'''
books_count = pd.read_sql(query_0, engine)
print(books_count)

#SQL query to find the number of books released after January 1, 2000
query_1 = f'''
SELECT COUNT(book_id) FROM books AS books_number
WHERE publication_date > '2000-01-01'
'''
books_count_after = pd.read_sql(query_1, engine)
print(books_count_after)

   count
0   1000
   count
0    819


<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

The total number of books is 1000. 
Most of the books were released after January 1, 2000 and their number is 819. 

#### 2. Find the number of user reviews and the average rating for each book.

<div class="alert alert-danger">
<s><b>Reviewer's comment: </b> Plase, re-read task and provide complete query.
</div>

<div class="alert alert-info"> <b>Student's comment:</b> Hope now I understood right</div>    

In [4]:
#SQL query to find the number of user reviews and the average rating for each book

query_2 = f''' SELECT COUNT(DISTINCT review_id) AS reviews_count, AVG(rating) AS avg_rating, books.book_id
FROM(( books
INNER JOIN ratings ON ratings.book_id = books.book_id)
INNER JOIN reviews ON reviews.book_id = ratings.book_id)
GROUP BY books.book_id
ORDER BY book_id ASC
'''

stat = pd.read_sql(query_2, engine)
display(stat)

reviews_count  avg_rating  book_id
0                2    3.666667        1
1                1    2.500000        2
2                3    4.666667        3
3                2    4.500000        4
4                4    4.000000        5
..             ...         ...      ...
989              3    3.666667      996
990              3    3.400000      997
991              4    3.200000      998
992              2    4.500000      999
993              4    3.833333     1000

[994 rows x 3 columns]

<div class="alert alert-danger">
<s><b>Reviewer's comment v2: </b>Good! Please, check your join key for reviews table.
</div>

<div class="alert alert-info"> <b>Student's comment:</b>Changed it from 'username' to 'book_id'</div>    

#### 3. Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [5]:
query_4 = f'''
SELECT publisher, COUNT(books.book_id)
FROM 
    publishers
INNER JOIN books ON books.publisher_id = publishers.publisher_id
WHERE num_pages > 50
GROUP BY publishers.publisher
ORDER BY COUNT(book_id) DESC
'''
publisher_name = pd.read_sql(query_4, engine)
display(publisher_name)

publisher  count
0                                        Penguin Books     42
1                                              Vintage     31
2                             Grand Central Publishing     25
3                                     Penguin Classics     24
4                                     Ballantine Books     19
..                                                 ...    ...
329                                         Turtleback      1
330  Atheneum Books for Young Readers: Richard Jack...      1
331                                     Penguin Signet      1
332                                    Victor Gollancz      1
333                      Harvard Business Review Press      1

[334 rows x 2 columns]

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

Top 3 Publishers that have released the greatest number of books with more than 50 pages are:
    - Penguin Books (published 42 books)
    - Vintage (published 31 books)
    - Grand Central Publishing (published 25 books)

#### 4. Identify the author with the highest average book rating: look only at books with at least 50 ratings.

In [6]:
# SQL query to identify the author with the highest average book rating
query_5 = f'''
SELECT author, AVG(rating)
FROM(( authors
INNER JOIN books ON books.author_id = authors.author_id)
INNER JOIN ratings ON ratings.book_id = books.book_id)
GROUP BY author
HAVING COUNT(ratings.book_id) > 49
ORDER BY AVG(rating) DESC
LIMIT 5
'''
top_author = pd.read_sql(query_5, engine)
display(top_author)

author       avg
0                     Diana Gabaldon  4.300000
1         J.K. Rowling/Mary GrandPré  4.288462
2                    Agatha Christie  4.283019
3  Markus Zusak/Cao Xuân Việt Khương  4.264151
4                     J.R.R. Tolkien  4.240964

Author Diana Gabaldon has the highest average book rating among the books having at least 50 (more than 49) ratings.
We shold pay attention to task description "at least 50 ratings", which is not equal to "more than 50" in the previous task.

In [7]:
#in case if we don't have to include 50 to the number of ratings the answer is J.K. Rowling/Mary GrandPré:

In [8]:
query_51 = f'''
SELECT author, AVG(rating)
FROM(( authors
INNER JOIN books ON books.author_id = authors.author_id)
INNER JOIN ratings ON ratings.book_id = books.book_id)
GROUP BY author
HAVING COUNT(ratings.book_id) > 50
ORDER BY AVG(rating) DESC
LIMIT 5
'''
top_author = pd.read_sql(query_51, engine)
display(top_author)

author       avg
0         J.K. Rowling/Mary GrandPré  4.288462
1                    Agatha Christie  4.283019
2  Markus Zusak/Cao Xuân Việt Khương  4.264151
3                     J.R.R. Tolkien  4.240964
4           Roald Dahl/Quentin Blake  4.209677

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

#### 5. Find the average number of text reviews among users who rated more than 50 books.

In [9]:
query_5 = f''' SELECT AVG(review_number) FROM 
(SELECT 
        COUNT(DISTINCT reviews.review_id) as review_number
    FROM 
        ratings
    INNER JOIN reviews ON reviews.username = ratings.username
    GROUP BY 
        reviews.username
    HAVING    
        COUNT(DISTINCT ratings.rating_id) > 50) as SUBQ
'''
top_author = pd.read_sql(query_5, engine)
display(top_author)

avg
0  24.333333

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

The average number of text reviews among users who rated more than 50 books is 24.3.

[Go back to the Table of Contents](#table_of_contents)